File: Week 01 — C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Prompt user for credit card number
    long number = get_long("Number: ");

    // Count the number of digits
    long temp = number;
    int digits = 0;
    while (temp > 0)
    {
        temp /= 10;
        digits++;
    }

    // Validate number of digits
    if (digits != 13 && digits != 15 && digits != 16)
    {
        printf("INVALID\n");
        return 0;
    }

    // Calculate checksum
    int sum = 0;
    temp = number;

    // Process digits from right to left
    for (int i = 0; i < digits; i++)
    {
        int digit = temp % 10;
        temp /= 10;

        if (i  % 2 == 1) // Every other digit
        {
            digit *= 2;
            sum += (digit % 10) + (digit / 10);
        }
        else // Other digits
        {
            sum += digit;
        }
    }

    // Check if checksum is valid
    if (sum % 10 != 0)
    {
        printf("INVALID\n");
        return 0;
    }

    // Determine card type based on starting digits
    long first_two = number;
    while (first_two >= 100)
    {
        first_two /= 10;
    }

    long first_one = first_two / 10;

    // Check American Express (15 digits, starts with 34 or 37)
    if (digits == 15 && (first_two == 34 || first_two == 37))
    {
        printf("AMEX\n");
    }
    // Check MasterCard (16 digits, starts with 51, 52, 53, 54, or 55)
    else if (digits == 16 && first_two >= 51 && first_two <= 55)
    {
        printf("MASTERCARD\n");
    }
    // Check Visa (13 or 16 digits, starts with 4)
    else if ((digits == 13 || digits == 16) && first_one == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: Week 06 — Python/sentimental-credit/credit.py
from cs50 import get_string
import re

def main():
    while True:
        try:
            number = get_string("Number: ")
            if number is None:
                return  # Handle EOF
            if re.match(r"^\d+$", number):
                break
        except EOFError:
            return

    if check_luhn(number):
        print_card_type(number)
    else:
        print("INVALID")

def check_luhn(number):
    sum = 0
    digits = [int(d) for d in number]
    
    # Iterate from right to left
    for i, digit in enumerate(reversed(digits)):
        if i % 2 == 1:
            product = digit * 2
            sum += (product // 10) + (product % 10)
        else:
            sum += digit
            
    return sum % 10 == 0

def print_card_type(number):
    length = len(number)
    
    if length == 15 and re.match(r"^3[47]", number):
        print("AMEX")
    elif length == 16 and re.match(r"^5[1-5]", number):
        print("MASTERCARD")
    elif (length == 13 or length == 16) and re.match(r"^4", number):
        print("VISA")
    else:
        print("INVALID")

if __name__ == "__main__":
    main()
